### PR TITLE
Get specific object version with S3 module.

### PR DESCRIPTION
**Pull request by wtcross**
This PR introduces the capability of specifying an object version id when using the `get` and `getstr` modes of the `S3 module.

### CODE REVIEW

**Comment by wtcross** [comment]
@bcoca Thanks! This is going to be a useful feature.

**Comment by wtcross** [comment]
I have tested both `get` and `getstr` in buckets that have versioning enabled as well as disabled. To enable versioning (if you would like to test yourself) go to an S3 bucket in the AWS Console, click Properties and then the Versioning dropdown. There's a big button to enable it.

**Comment by wtcross** [comment]
bucket-versioning

**Comment by wtcross** [comment]
For reference: http://docs.aws.amazon.com/AmazonS3/latest/dev/ObjectVersioning.html

**Comment by wtcross** [comment]
@bennojoy I resolved the merge conflict by rebasing `devel`. All tests are passing

**Comment by wtcross** [comment]
rebased `devel` again

Is there a specific order in which PRs are evaluated after being assigned a status? I want to write a blog post about using the `S3` module on my companies blog, but need this feature before doing so.

**Comment by bcoca** [comment]
normally by date, ease of integration and sometimes random pings from users reminding us that this is almost done

File: cloud/amazon/s3.py [hunk 1]
```diff
@@ -73,6 +73,12 @@
       - Keyname of the object inside the bucket. Can be used to create "virtual directories", see examples.
     required: false
     default: null
+  version:
```
Comment (bcoca) [inline cloud/amazon/s3.py:76]: still requries version_added

**Comment by bcoca** [comment]
@bennojoy test functionally when you can?

File: cloud/amazon/s3.py (hunk 1 above)
Comment (wtcross) [inline cloud/amazon/s3.py:76]: Good catch! I'll add it now.

**Comment by bennojoy** [comment]
@wtcross  could you please change the version added from 1.9 to 2.0 also the put operation doesn't seem to support version, any reason we have not aded version for put operation ?

**Comment by wtcross** [comment]
@bennojoy  I rewrote my last commit to have the correct version.

Updating a specific version of an object would break the semantics of versioning. In any case, the [`boto.s3.key.Key#set_contents_from_file`](http://boto.readthedocs.org/en/latest/ref/s3.html#boto.s3.key.Key.set_contents_from_file) method does have a `version_id` parameter. Only read or metadata-related operations have the `version_id` parameter. These are the methods of the `boto.s3.key.Key` class that the `s3` module uses:
- [`get_contents_to_filename`](http://boto.readthedocs.org/en/latest/ref/s3.html#boto.s3.key.Key.get_contents_to_filename)
- [`get_contents_as_string`](http://boto.readthedocs.org/en/latest/ref/s3.html#boto.s3.key.Key.get_contents_as_string)

Here's a good image showing what happens for a versioned object:
version diagram

Updates go on top and reads can specify any version.

**Comment by wtcross** [comment]
@bennojoy I made sure that all my changes are in one commit, so that I'm not polluting Ansible commit history. Please let me know if there is anything else you need me to do and I will.

**Comment by wtcross** [comment]
By the way, I apologize for not using the issue template. Future PRs will definitely use that.

**Comment by wtcross** [comment]
@bennojoy Do you have any other questions or concerns?

**Comment by bennojoy** [comment]
@wtcross  all good will just give it a try on my end and merge

**Comment by bennojoy** [comment]
Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:

https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code
Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Pull request merged**